akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.io.dns.internal
+
+import java.net.{ InetAddress, InetSocketAddress }
+
+import akka.actor.{ Actor, ActorLogging, ActorRef, NoSerializationVerificationNeeded, Stash }
+import akka.io.dns.{ DnsRecordClass, DnsRecordType, DnsResourceRecord }
+import akka.io.{ IO, Udp }
+
+import scala.collection.immutable.Seq
+import scala.collection.immutable
+
+object DnsClient {
+  case class SrvQuestion(id: Short, name: String)
+
+  case class Question4(id: Short, name: String)
+
+  case class Question6(id: Short, name: String)
+
+  case class Answer(id: Short, rrs: immutable.Seq[DnsResourceRecord]) extends NoSerializationVerificationNeeded
+
+}
+
+class DnsClient(ns: InetSocketAddress) extends Actor with ActorLogging with Stash {
+
+  import DnsClient._
+
+  import context.system
+
+  IO(Udp) ! Udp.Bind(self, new InetSocketAddress(InetAddress.getByAddress(Array.ofDim(4)), 0))
+
+  var requests: Map[Short, ActorRef] = Map.empty
+
+  def receive: Receive = {
+    case Udp.Bound(local) ⇒
+      log.debug(s"Bound to UDP address $local")
+      context.become(ready(sender()))
+      unstashAll()
+    case _: Question4 ⇒
+      stash()
+    case _: Question6 ⇒
+      stash()
+    case _: SrvQuestion ⇒
+      stash()
+  }
+
+  private def message(name: String, id: Short, recordType: DnsRecordType): Message = {
+    Message(
+      id,
+      MessageFlags(),
+      Seq(
+        DnsQuestion(name, recordType, DnsRecordClass.IN)
+      ))
+  }
+
+  def ready(socket: ActorRef): Receive = {
+    {
+      case Question4(id, name) ⇒
+        log.debug(s"Resolving $name (A)")
+        requests += (id -> sender())
+        log.debug(s"Message to $ns: ${message(name, id, DnsRecordType.A)}")
+        socket ! Udp.Send(message(name, id, DnsRecordType.A).write(), ns)
```

Comment:
What happens if the UPD messages are dropped and we don't get any reply? Do we retry or is that rare enough that it's enough to timeout?

Reply:
Perhaps we should retry on a different name server?

Reply:
After doing some research i've changed this to try the name servers one at a time. If a timeout occurs it moves to the next